Orange/widgets/utils/itemmodels.py

```diff
@@ -265,15 +265,21 @@ def __delitem__(self, i):
         self.endRemoveRows()
 
     def __setitem__(self, i, value):
+        self._check_sort_order()
         if isinstance(i, slice):
+            self.beginResetModel()
             start, stop, _ = _as_contiguous_range(i, len(self))
-            stop -= 1
+            self.removeRows(start, stop - start)
+            self.insertRows(start, len(value))
+            self._table[start:start + len(value)] = value
+            self.dataChanged.emit(
+                self.index(start, 0),
+                self.index(start + len(value) - 1, self.columnCount() - 1))
+            self.endResetModel()
```

Comment:
@ales-erjavec, this code now calls `removeRows` and `insertRows` to properly handle `_roledata`, too. These two methods emit (begin|end)(Remove|Insert)Rows. Afterwards it modifies the table and emits `dataChanged`.

But still: if I connect Dataset (with Philadelphia crime) to Aggregate (from Time Series add-on), then remove the connection and connect it again, the view in Aggregate remains empty -- unless I call `beginResetModel` and `endResetModel` . After a day of exploration, I've no clue why. I see that the view does call `rowCount` and `columnCount` (which give correct results), all model's data seems properly initialized, but the view is empty.

Do you see why is this, or should I just remove `dataChanged` and we reset the model? In practice, I doubt it will make a difference because this code is almost always called when we replace the entire model's data (`self.model[:] = ...`).

Reply:
The problem was that `columnsInserted|Removed` were not emitted. The model was cleared (`clear()`) which reset its column count to 0, then when the model was populated with `model[:] = ...` it only ever emitted `rowsInserted`, never `columnsInserted` event though the models `columnCount()` changed  implicitly.